protosanitizer/protosanitizer.go

```diff
@@ -48,67 +48,68 @@ type stripSecrets struct {
 }
 
 func (s *stripSecrets) String() string {
-	// First convert to a generic representation. That's less efficient
-	// than using reflect directly, but easier to work with.
-	var parsed interface{}
-	b, err := json.Marshal(s.msg)
-	if err != nil {
-		return fmt.Sprintf("<<json.Marshal %T: %s>>", s.msg, err)
-	}
+	stripped := s.msg
+
+	// also support scalar types like string, int, etc.
 	msg, ok := s.msg.(proto.Message)
-	if !ok {
-		return string(b)
-	}
-	if err := json.Unmarshal(b, &parsed); err != nil {
-		return fmt.Sprintf("<<json.Unmarshal %T: %s>>", s.msg, err)
+	if ok {
+		stripped = stripMessage(msg.ProtoReflect())
 	}
 
-	// Now remove secrets from the generic representation of the message.
-	s.strip(parsed, msg.ProtoReflect())
-
-	// Re-encoded the stripped representation and return that.
-	b, err = json.Marshal(parsed)
+	b, err := json.Marshal(stripped)
 	if err != nil {
 		return fmt.Sprintf("<<json.Marshal %T: %s>>", s.msg, err)
 	}
 	return string(b)
 }
 
-func (s *stripSecrets) strip(parsed interface{}, msg protoreflect.Message) {
-	// The corresponding map in the parsed JSON representation.
-	parsedFields, ok := parsed.(map[string]interface{})
-	if !ok {
-		// Probably nil.
-		return
+func stripSingleValue(field protoreflect.FieldDescriptor, v protoreflect.Value) any {
+	switch field.Kind() {
+	case protoreflect.MessageKind:
+		return stripMessage(v.Message())
+	case protoreflect.EnumKind:
+		return field.Enum().Values().ByNumber(v.Enum()).Name()
+	default:
+		return v.Interface()
 	}
+}
+
+func stripValue(field protoreflect.FieldDescriptor, v protoreflect.Value) any {
+	if field.IsList() {
+		l := v.List()
+		res := make([]any, l.Len())
+		for i := range l.Len() {
+			res[i] = stripSingleValue(field, l.Get(i))
```

Comment:
For example, what happens when a list has a list? Will it call `stripSingleValue()` with a list, which then leads to `default:` branch there and thus print the list with potential secrets?

Reply:
List is represented as repeated field in the protobuf. Like:
```proto3
  repeated VolumeCapability volume_capabilities = 3;
```
And it cannot be nested without going through a level of message. i.e.:
```proto3
  repeated repeated VolumeCapability volume_capabilities = 3;
```
is not valid, but this one is:
```proto3
  repeated Topology accessible_topology = 5;

message Topology {
  map<string, string> segments = 1;
}
```

Similarly, `map<string, map<string, string>>` or `repeated map<string, string>` are not valid.

Reply:
Thanks for the explanation!
/lgtm
/approve